File: python_bindings_jupedsim/agent.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
#include "GeneralizedCentrifugalForceModel.hpp"
#include "GenericAgent.hpp"
#include "conversion.hpp"

#include <pybind11/cast.h>
#include <pybind11/detail/common.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h> // IWYU pragma: keep

#include <cstdint>
#include <tuple>

namespace py = pybind11;

void init_agent(py::module_& m)
{
    py::class_<GenericAgent>(m, "Agent")
        .def(
            py::init([](uint64_t journeyId,
                        uint64_t stageId,
                        std::tuple<double, double> position,
                        std::tuple<double, double> orientation,
                        GenericAgent::Model model,
                        double fatigue,
                        double fatigue_rate,
                        std::string type,
                        double heartbeat,
                        double stress,
                        double panic,
                        double openness,
                        double conscientiousness,
                        double extraversion,
                        double agreeableness,
                        double neuroticism) {
                auto agent = GenericAgent(
                    GenericAgent::ID::Invalid,
                    journeyId,
                    stageId,
                    intoPoint(position),
                    intoPoint(orientation),
                    model);
                agent.fatigue = fatigue;
                agent.fatigue_rate = fatigue_rate;
                agent.type = type;
                agent.heartbeat = heartbeat;
                agent.stress = stress;
                agent.panic = panic;
                agent.openness = openness;
                agent.conscientiousness = conscientiousness;
                agent.extraversion = extraversion;
                agent.agreeableness = agreeableness;
                agent.neuroticism = neuroticism;
                return agent;
            }),
            py::kw_only(),
            py::arg("journey_id"),
            py::arg("stage_id"),
            py::arg("position"),
            py::arg("orientation"),
            py::arg("model"),
            py::arg("fatigue") = 0.0,
            py::arg("fatigue_rate") = 0.0,
            py::arg("type") = "male",
            py::arg("heartbeat") = 70.0,
            py::arg("stress") = 0.0,
            py::arg("panic") = 0.0,
            py::arg("openness") = 0.5,
            py::arg("conscientiousness") = 0.5,
            py::arg("extraversion") = 0.5,
            py::arg("agreeableness") = 0.5,
            py::arg("neuroticism") = 0.5)
        .def_property_readonly("id", [](const GenericAgent& agent) { return agent.id.getID(); })
        .def_property_readonly(
            "journey_id", [](const GenericAgent& agent) { return agent.journeyId.getID(); })
        .def_property_readonly(
            "stage_id", [](const GenericAgent& agent) { return agent.stageId.getID(); })
        .def_property_readonly(
            "position", [](const GenericAgent& agent) { return intoTuple(agent.pos); })
        .def_property_readonly(
            "orientation", [](const GenericAgent& agent) { return intoTuple(agent.orientation); })
        .def_property(
            "target",
            [](const GenericAgent& agent) { return intoTuple(agent.target); },
            [](GenericAgent& agent, std::tuple<double, double> target) {
                agent.target = intoPoint(target);
            })
        .def_property(
            "fatigue",
            [](const GenericAgent& agent) { return agent.fatigue; },
            [](GenericAgent& agent, double fatigue) { agent.fatigue = fatigue; })
        .def_property(
            "fatigue_rate",
            [](const GenericAgent& agent) { return agent.fatigue_rate; },
            [](GenericAgent& agent, double fatigue_rate) { agent.fatigue_rate = fatigue_rate; })
        .def_property(
            "type",
            [](const GenericAgent& agent) { return agent.type; },
            [](GenericAgent& agent, std::string type) { agent.type = type; })
        .def_property(
            "heartbeat",
            [](const GenericAgent& agent) { return agent.heartbeat; },
            [](GenericAgent& agent, double heartbeat) { agent.heartbeat = heartbeat; })
        .def_property(
            "stress",
            [](const GenericAgent& agent) { return agent.stress; },
            [](GenericAgent& agent, double stress) { agent.stress = stress; })
        .def_property(
            "panic",
            [](const GenericAgent& agent) { return agent.panic; },
            [](GenericAgent& agent, double panic) { agent.panic = panic; })
        .def_property(
            "openness",
            [](const GenericAgent& agent) { return agent.openness; },
            [](GenericAgent& agent, double val) { agent.openness = val; })
        .def_property(
            "conscientiousness",
            [](const GenericAgent& agent) { return agent.conscientiousness; },
            [](GenericAgent& agent, double val) { agent.conscientiousness = val; })
        .def_property(
            "extraversion",
            [](const GenericAgent& agent) { return agent.extraversion; },
            [](GenericAgent& agent, double val) { agent.extraversion = val; })
        .def_property(
            "agreeableness",
            [](const GenericAgent& agent) { return agent.agreeableness; },
            [](GenericAgent& agent, double val) { agent.agreeableness = val; })
        .def_property(
            "neuroticism",
            [](const GenericAgent& agent) { return agent.neuroticism; },
            [](GenericAgent& agent, double val) { agent.neuroticism = val; })
        .def_property_readonly(
            "model",
            [](GenericAgent& agent) -> auto& { return agent.model; },
            py::return_value_policy::reference);
}

File: python_modules/jupedsim/jupedsim/agent.py
# SPDX-License-Identifier: LGPL-3.0-or-later

import jupedsim.native as py_jps
from jupedsim.models.anticipation_velocity_model import (
    AnticipationVelocityModelState,
)
from jupedsim.models.collision_free_speed import CollisionFreeSpeedModelState
from jupedsim.models.collision_free_speed_v2 import (
    CollisionFreeSpeedModelV2State,
)
from jupedsim.models.generalized_centrifugal_force import (
    GeneralizedCentrifugalForceModelState,
)
from jupedsim.models.social_force import SocialForceModelState


class Agent:
    """Represents an Agent in the simulation.

    Agent objects are always retrieved from the simulation and never created directly.

    Agents can be accessed with:

    .. code:: python

        # a specific agent
        sim.agent(id)

        # all agents as iterator
        sim.agents()

        # agents in a specific distance to a point as iterator
        sim.agents_in_range(position, distance)

        # agents in a polygon as iterator
        sim.agents_in_polygon(polygon)

    .. note ::

        You need to be aware that currently there are no checks done when setting
        properties on an Agent instance. For example it is possible to set an Agent position
        outside the walkable area of the Simulation resulting in a crash.
    """

    def __init__(self, backing) -> None:
        """Do not use.

        Retrieve agents from the simulation.
        """
        self._obj = backing

    @property
    def id(self) -> int:
        """Numeric id of the agent in this simulation."""
        return self._obj.id

    @property
    def journey_id(self) -> int:
        """Id of the :class:`~jupedsim.journey.JourneyDescription` the agent is currently following."""
        return self._obj.journey_id

    @property
    def stage_id(self) -> int:
        """Id of the :class:`Stage` the Agent is currently targeting."""
        return self._obj.stage_id

    @property
    def position(self) -> tuple[float, float]:
        """Position of the agent."""
        return self._obj.position

    @property
    def orientation(self) -> tuple[float, float]:
        """Orientation of the agent."""
        return self._obj.orientation

    @property
    def target(self) -> tuple[float, float]:
        """Current target of the agent.

        Can be used to directly steer an agent towards the given coordinate.
        This will bypass the strategical and tactical level, but the operational level
        will still be active.

        .. important::

            If the agent is not in a Journey with a DirectSteering stage, any change will be
            ignored.

        .. important::

            When setting the target, the given coordinates must lie within the walkable area.
            Otherwise, an error will be thrown at the next iteration call.

        Returns:
            Current target of the agent.
        """
        return self._obj.target

    @target.setter
    def target(self, target: tuple[float, float]):
        self._obj.target = target

    @property
    def type(self) -> str:
        """Type of the agent (e.g., 'male', 'female', 'child')."""
        return self._obj.type

    @type.setter
    def type(self, type: str):
        self._obj.type = type

    @property
    def heartbeat(self) -> float:
        """Heartbeat of the agent in bpm."""
        return self._obj.heartbeat

    @heartbeat.setter
    def heartbeat(self, val: float):
        self._obj.heartbeat = val

    @property
    def stress(self) -> float:
        """Stress level of the agent (0.0 to 1.0)."""
        return self._obj.stress

    @stress.setter
    def stress(self, val: float):
        self._obj.stress = val

    @property
    def panic(self) -> float:
        """Panic level of the agent (0.0 to 1.0)."""
        return self._obj.panic

    @panic.setter
    def panic(self, val: float):
        self._obj.panic = val

    @property
    def openness(self) -> float:
        """Openness trait of the agent (0.0 to 1.0)."""
        return self._obj.openness

    @openness.setter
    def openness(self, val: float):
        self._obj.openness = val

    @property
    def conscientiousness(self) -> float:
        """Conscientiousness trait of the agent (0.0 to 1.0)."""
        return self._obj.conscientiousness

    @conscientiousness.setter
    def conscientiousness(self, val: float):
        self._obj.conscientiousness = val

    @property
    def extraversion(self) -> float:
        """Extraversion trait of the agent (0.0 to 1.0)."""
        return self._obj.extraversion

    @extraversion.setter
    def extraversion(self, val: float):
        self._obj.extraversion = val

    @property
    def agreeableness(self) -> float:
        """Agreeableness trait of the agent (0.0 to 1.0)."""
        return self._obj.agreeableness

    @agreeableness.setter
    def agreeableness(self, val: float):
        self._obj.agreeableness = val

    @property
    def neuroticism(self) -> float:
        """Neuroticism trait of the agent (0.0 to 1.0)."""
        return self._obj.neuroticism

    @neuroticism.setter
    def neuroticism(self, val: float):
        self._obj.neuroticism = val

    @property
    def model(
        self,
    ) -> (
        GeneralizedCentrifugalForceModelState
        | CollisionFreeSpeedModelState
        | CollisionFreeSpeedModelV2State
        | AnticipationVelocityModelState
        | SocialForceModelState
    ):
        """Access model specific state of this agent."""
        model = self._obj.model
        if isinstance(model, py_jps.GeneralizedCentrifugalForceModelState):
            return GeneralizedCentrifugalForceModelState(model)
        elif isinstance(model, py_jps.CollisionFreeSpeedModelState):
            return CollisionFreeSpeedModelState(model)
        elif isinstance(model, py_jps.CollisionFreeSpeedModelV2State):
            return CollisionFreeSpeedModelV2State(model)
        elif isinstance(model, py_jps.AnticipationVelocityModelState):
            return AnticipationVelocityModelState(model)
        elif isinstance(model, py_jps.SocialForceModelState):
            return SocialForceModelState(model)
        else:
            raise Exception("Internal error")
